src/openms/include/OpenMS/ANALYSIS/ID/FalseDiscoveryRate.h

```diff
@@ -114,6 +158,488 @@ namespace OpenMS
+    //TODO even better: store nr targets and nr decoys when creating the groups!
+    //TODO alternative scoring is possible, too (e.g. ratio of tgts vs decoys),
+    // this requires templatization of the scores_labels vector though
+    void getScores_(
+        std::vector<std::pair<double,bool>>& scores_labels,
+        const std::vector<ProteinIdentification::ProteinGroup>& grps,
+        const std::unordered_set<std::string>& decoy_accs) const;
+
+    template<class ...Args>
+    void getPeptideScoresFromMap_(
+        std::vector<std::pair<double,bool>>& scores_labels,
+        const ConsensusMap & cmap, Args&& ... args) const
+    {
+      std::function<void (const PeptideIdentification &)> f =
+          [&, this](const PeptideIdentification& id) -> void {this->getScores_(scores_labels, id, std::forward<Args>(args)...);};
+      cmap.applyFunctionOnPeptideIDs(f);
+    }
+
+    void getScores_(
+        std::vector<std::pair<double,bool>>& scores_labels,
+        const ProteinIdentification & id) const
+    {
+
+      scores_labels.reserve(scores_labels.size() + id.getHits().size());
+      std::transform(id.getHits().begin(), id.getHits().end(),
+                     std::back_inserter(scores_labels),
+                     [](const ProteinHit& hit)
+                     {
+                       checkTDAnnotation_(hit);
+                       return std::make_pair<double,bool>(hit.getScore(), getTDLabel_(hit));
+                     }
+                    );
+
+    }
+
+    void getScores_(
+        std::vector<std::pair<double,bool>>& scores_labels,
+        const PeptideIdentification & id, bool all_hits, int charge, const String& identifier) const
+    {
+      if (id.getIdentifier() == identifier)
+      {
+        getScores_(scores_labels, id, all_hits, charge);
+      }
+    }
+
+
+    void getScores_(
+        std::vector<std::pair<double,bool>>& scores_labels,
+        const PeptideIdentification & id, bool all_hits, const String& identifier) const
+    {
+      if (id.getIdentifier() == identifier)
+      {
+        getScores_(scores_labels, id, all_hits);
+      }
+    }
+
+    void getScores_(
+        std::vector<std::pair<double,bool>>& scores_labels,
+        const PeptideIdentification & id, int charge, const String& identifier) const
+    {
+      if (id.getIdentifier() == identifier)
+      {
+        getScores_(scores_labels, id, charge);
+      }
+    }
+
+    template<typename IDType, typename std::enable_if<IsIDType<IDType>::value>::type* = nullptr>
+    void getScores_(
+        std::vector<std::pair<double,bool>>& scores_labels,
+        const IDType & id, const String& identifier) const
+    {
+      if (id.getIdentifier() == identifier)
+      {
+        getScores_(scores_labels, id);
+      }
+    }
+
+    static bool getTDLabel_(const MetaInfoInterface& idOrHit)
+    {
+      return std::string(idOrHit.getMetaValue("target_decoy"))[0] == 't';
+    }
+
+    template<class ...Args>
+    void getScores_(
```

Comment:
can all of the template metaprogramming go into its own .h file? it really blows up the current file and makes it hard to read

Reply:
the header file goes from 120 LOC to 700 LOC...